docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,777 @@
+}
+
+type ConnOpenConfirm struct {
+  local             ConnectionParty
+  remoteStateProof  Proof
+}
+```
+
+Notice that the fields in each datagram match with the corresponding protocol handler signature.
+
+##### Relayer Logic
+
+A correct relayer periodically scans a pair of chains and, if the chains are in a specific state (e.g., a connection is `INIT` on one of the chains and inexistent in the other chain), then the relayer performs the necessary steps in the ICS3 protocol (e.g., construct the necessary `ConnOpenTry` datagram and pushes this datagram to the second chain).
+
+
+###### Step 2.a:
+
+```golang
+// Assume `cAlice` is a connection on chain Alice.
+if (cAlice.state == INIT) {
+  // Handshake started at Alice (step 1 done because the connection is INIT).
+
+  // Fetch the connection end from the other chain (Bob)
+  cBob := chains[Bob].getConnection(cAlice.remote.connectionIdentifier)
+  
+  // We will do step 2.a, by pushing a `ConnOpenTry` to Bob.
+  if (cBob == null || cBob.state == INIT) {
+    // Construct the appropriate datagram.
+    d := ConnOpenTry{
+      cAlice.remote,                // Bob's end (a ConnectionParty object)
+      cAlice.local,                 // Alice's end
+      cAlice.version,               // Alice's proposed version
+      constructStateProof(cAlice),
+      constructClientProof(cAlice)}
+
+    // Push the datagram to Bob's module
+    chains[Bob].push(d)
+  }
+}
+```
+
+###### Step 3.a:
+
+```golang
+// Assume `cBob` is a connection on chain Bob.
+if (cBob.state == TRYOPEN) {
+  // Fetch the matching connection end from Alice's side
+  cAlice := chains[Alice].getConnection(cBob.remote.connectionIdentifier)
+  
+  // Handshake ConnOpenTry (step 2) done at Bob.
+  // We will push a `ConnOpenAck` to Alice. This is step 3.a.
+  if (cAlice.state == INIT || cAlice.state == TRYOPEN) {
+    // Construct the appropriate datagram.
+    d := ConnOpenAck{
+      cBob.remote,                // Alice's party details
+      cBob.version,               // Bob's proposed version
+      constructStateProof(cBob),
+      constructClientProof(cBob)}
+
+    // Push the datagram to Bob's module
+    chains[Alice].push(d)
+  }
+}
+```
+
+###### Step 4.a:
+
+```golang
+// Assume `cAlice` is a connection on chain Alice.
+if (cAlice.state == OPEN) {
+  // Fetch the matching connection end from Alice's side
+  cBob := chains[Bob].getConnection(cAlice.remote.connectionIdentifier)
+  
+  // Handshake ConnOpenAck (step 3) done at Alice.
+  // We will push a `ConnOpenConfirm` to Bob. This is step 4.a.
+  if (cBob.state == TRYOPEN) {
+    // Construct the appropriate datagram.
+    d := ConnOpenConfirm{
+      cBob.local,                 // Bob's party details
+      constructStateProof(cAlice)}
+
+    // Push the datagram to Bob's module
+    chains[Bob].push(d)
+  }
+}
+```
+
+#### Helper functions & application-level predicates
+
+###### `verifyProofs`
+
+```golang
+func verifyProofs(remoteStateProof Proof, 
```

Comment:
I would make this function call a `verifyProof` for all proofs. Or even do two `verifyProof` calls in the callers/ handlers.
Then a `verifyProof` will do both what you call syntactical and semantical checks